content/blog/container-filesystems/index.md

```diff
@@ -0,0 +1,132 @@
+
+No matter the reason, in this post, we’ll cover a few methods you can use to inspect the files inside a container.
+
+We’ll start with the easy and commonly recommended ways of exploring a container’s filesystem, and talk about why they don’t always work. We’ll follow that up with a basic understanding of how container filesystems are managed by the Linux kernel, and we’ll use that understanding to inspect the filesystem in different, but still easy, ways.
+
+## Method 1: Exec into the container
+
+If you perform a quick search on how to inspect a container’s filesystem, a common solution you’ll find is to use the [Docker command](https://docs.docker.com/engine/reference/commandline/exec/) ([[1]](https://stackoverflow.com/questions/20813486/exploring-docker-containers-file-system), [[2]](https://www.baeldung.com/ops/docker-container-filesystem)):
+
+```
+docker exec -it mycontainer /bin/bash
+```
+
+This is a great way to start. And if it works for all your needs, you should continue using it.
+
+One downside of this approach, however, is that it requires a shell to be present inside the container. If no `/bin/bash`, `/bin/sh` or other shell is present inside the container, then this approach won’t work. Many of the containers we build for the Pixie project, for example, are based on `distroless` and don’t have a shell included to keep image sizes small. In those cases, this approach doesn’t work.
+
+Even if a shell is available, you won’t have access to all the tools you’re used to. So if there’s no `grep` installed inside the container, then you also won’t have access to `grep`. That’s another reason to look for something better.
+
+## Method 2: Using nsenter
+
+If you get a little more advanced, you’ll realize that container processes are just like other processes on the Linux host, only running inside a namespace to keep them isolated from the rest of the system.
+
+So you could use the [`nsenter`](https://man7.org/linux/man-pages/man1/nsenter.1.html) command to enter the namespace of the target container, using something like this:
+
+```
+# Get the host PID of the process in the container
+PID=$(docker container inspect mycontainer | jq '.[0].State.Pid')
+
+# Use nsenter to go into the container’s mount namespace.
+sudo nsenter -m -t $PID /bin/bash
+```
+
+This enters the mount (`-m`) namespace of the target process (`-t $PID`), and runs `/bin/bash`. Entering the mount namespace essentially means we get the view of the filesystem that the container sees.
+
+This approach may seem more promising than the `docker exec` approach, but runs into a similar issue: it requires `/bin/bash` (or some other shell) to be in the target container. If we were entering anything other than the mount namespace, we could still access the files on the host, but because we’re entering the mount namespace before executing `/bin/bash` (or other shell), we’re out of luck if there’s no shell inside the mount namespace.
+
```

Comment:
I feel that docker cp probably worth mentioning, if the file paths were know already.
https://docs.docker.com/engine/reference/commandline/cp/

Another interesting approach is docker create & docker export:
https://stackoverflow.com/a/53481010